### DISCUSSION TITLE
Naive question about unknown-var test case

### DISCUSSION THREAD

**Discussion by nobodxbodon** [aya-prover/aya-dev#75] [Q&A] [answered]
In unknown-var.aya:
```
def Nat : Type | zero
def p => Nat::suc Nat::zero
```
Checking it outputs as expected below:

> Error: The qualified name `Nat::suc` is not defined in the current scope

Question is, after I remove `Nat::suc`, it still gives similar error about `Nat::zero`:

> Error: The qualified name `Nat::zero` is not defined in the current scope

In module-basic.aya, it seems ::zero works. So why not in unknown-var.aya?

**Comment by ice1000** [comment, marked as answer]
Did you mean `data Nat`?

> In module-basic.aya, it seems ::zero works.

Because they did `open data Nat` instead of `data Nat`. The keyword `open` brings `Nat::*` into the current scope. You can think of it as `using namespace` in C++.

**Reply by ice1000** [reply]
I think the test case might have a typo. I'll go fix it.

Btw @nobodxbodon, I saw you commented the initial commit:
![image](https://user-images.githubusercontent.com/16398479/136648002-2e0dedad-1494-4177-993d-8cec79653b9b.png)

Could you please send me a screenshot of your comment? I opened the notification (https://github.com/aya-prover/aya-dev/commit/de31e4fbaed2c699ee56b068e484abaa42e89199) and I cannot find a comment in this webpage.

**Reply by nobodxbodon** [reply]
Yes the comment is about the typo
<img width="810" alt="Screen Shot 2021-10-09 at 12 08 47 AM" src="https://user-images.githubusercontent.com/392497/136648142-70e834bc-9b92-4482-8674-44bd3354b9bc.png">
:

**Reply by ice1000** [reply]
It's not a typo.

**Reply by nobodxbodon** [reply]
> Did you mean `data Nat`?
> 
> > In module-basic.aya, it seems ::zero works.
> 
> Because they did `open data Nat` instead of `data Nat`. The keyword `open` brings `Nat::*` into the current scope. You can think of it as `using namespace` in C++.

Thanks. The way unknown-var.aya is written now seems to me like "Nat::zero" is the right way to reference a known variable, thus the confusion.

**Reply by nobodxbodon** [reply]
> It's not a typo.

Is it meant to be distinguished with "zero"? I wonder why?

**Reply by ice1000** [reply]
The right way to reference a known (which means either imported or defined in the current file) variable is to use its name if it's in scope, and use its qualified name otherwise.

**Reply by ice1000** [reply]
> Is it meant to be distinguished with "zero"?

Yes

> I wonder why?

Yesterday you told me 'bout the blue blue sky🎶

**Reply by ice1000** [reply]
I think the point of unknown-var.aya is to test the error message of an unknown qualified name, and that's why there's no `suc` defined -- we are using an undefined variable, and we expect an error message about unknown qualified name.

Anyways, why are you learning from some wrong code? I found it very interesting while quite confusing.

**Reply by nobodxbodon** [reply]
I wanted to see what errors can be found, so I started from operator/ambiguous.aya in failure folder. Then I wanted to try Chinese identifiers, so I tried in unknown-var first and then rec-adt.aya (glad to see renaming `negs` to `负` works). 

I'm a bit surprised that you're surprised :) I find the test cases for errors are at least as interesting as the functional ones.

> I think the point of unknown-var.aya is to test the error message of an unknown qualified name, and that's why there's no suc defined -- we are using an undefined variable, and we expect an error message about unknown qualified name.

For this purpose, can unknown-var.aya be simplified to below? 
```
def Nat : Type
def p => Nat::suc
```

**Reply by nobodxbodon** [reply]
Saw [your fix](https://github.com/aya-prover/aya-dev/commit/4e548f1ac5d15704c9e926306582e7eb22689c83). Looks good.

**Reply by ice1000** [reply]
We even support emoji identifiers. But I don't think it's a good idea. This might be changed.

**Reply by ice1000** [reply]
I don't know. I don't have a persuasive argument.
